src/network/lan.cpp

```diff
@@ -0,0 +1,335 @@
+
+		struct addrinfo hints
+		{
+		};
+
+		hints.ai_family = AF_INET6;
+		hints.ai_socktype = SOCK_DGRAM;
+		hints.ai_flags = AI_V4MAPPED | AI_ADDRCONFIG;
+		struct addrinfo *result = nullptr;
+		if (!getaddrinfo("ff02::1", nullptr, &hints, &result)) {
+			for (auto info = result; info; info = info->ai_next) {
+				try {
+					sockaddr_in6 addr = *((struct sockaddr_in6 *)info->ai_addr);
+					addr.sin6_port = htons(adv_port);
+					UDPSocket socket_send(true);
+					int set_option_on = 1;
+					//setsockopt(socket_send.GetHandle(), SOL_SOCKET, IPV6_MULTICAST_HOPS,
+					//		(const char *)&set_option_on, sizeof(set_option_on));
+					auto use_scopes = scopes;
+					if (addr.sin6_scope_id) {
+						use_scopes.clear();
+						use_scopes.push_back(addr.sin6_scope_id);
+					}
+					for (auto &scope : use_scopes) {
+						addr.sin6_scope_id = scope;
+						socket_send.Send(Address(addr), str.c_str(), str.size());
+					}
+				} catch (const std::exception &e) {
+					verbosestream << "udp broadcast send over ipv6 fail " << e.what() << "\n";
+				}
+			}
+			freeaddrinfo(result);
+		}
+	}
+
+	try {
+		sockaddr_in addr = {};
+		addr.sin_family = AF_INET;
+		addr.sin_port = htons(adv_port);
+		UDPSocket socket_send(false);
+
+		// Full discloser, the use of inet_proto and the mreq stucture 
+		// were suggested by ai.
+
+		inet_pton(AF_INET, adv_multicast_addr, &(addr.sin_addr));
+
+		struct ip_mreq mreq;
+
+		mreq.imr_multiaddr.s_addr = inet_addr(adv_multicast_addr);
+		mreq.imr_interface.s_addr = htonl(INADDR_ANY);
+		
+
+		setsockopt(socket_send.GetHandle(), IPPROTO_IP, IP_ADD_MEMBERSHIP,
+				(const char *)&mreq, sizeof(mreq));
+
+		//int set_option_on = 2;
+		//setsockopt(socket_send.GetHandle(), SOL_SOCKET, IP_MULTICAST_TTL,
+		//		(const char *)&set_option_on, sizeof(set_option_on));
+
+		socket_send.Send(Address(addr), str.c_str(), str.size());
+	} catch (const std::exception &e) {
+		verbosestream << "udp broadcast send over ipv4 fail " << e.what() << "\n";
+	}
+}
+
+void lan_adv::serve(unsigned short port)
+{
+	server_port = port;
+	stop();
+	start(); 
+}
+
+void *lan_adv::run()
+{
+	BEGIN_DEBUG_EXCEPTION_HANDLER;
+
+	setName("lan_adv " + (server_port ? std::string("server") : std::string("client")));
+	UDPSocket socket_recv(g_settings->getBool("enable_ipv6"));
+
+	int set_option_off = 0, set_option_on = 1;
+	setsockopt(socket_recv.GetHandle(), SOL_SOCKET, SO_REUSEADDR,
+			(const char *)&set_option_on, sizeof(set_option_on));
+#ifdef SO_REUSEPORT
+	setsockopt(socket_recv.GetHandle(), SOL_SOCKET, SO_REUSEPORT,
+			(const char *)&set_option_on, sizeof(set_option_on));
+#endif
+	struct ip_mreq mreq;
+
+	mreq.imr_multiaddr.s_addr = inet_addr(adv_multicast_addr);
+	mreq.imr_interface.s_addr = htonl(INADDR_ANY);
+
+	setsockopt(socket_recv.GetHandle(), IPPROTO_IP, IP_ADD_MEMBERSHIP,
+			(const char *)&mreq, sizeof(mreq));
+	setsockopt(socket_recv.GetHandle(), IPPROTO_IPV6, IPV6_V6ONLY,
+			(const char *)&set_option_off, sizeof(set_option_off));
```

Comment:
more code that could be moved to the socket class.

Reply:
I don't think the lines setting the socket options for reuse address and reuse port need to be moved. Sense they are not necessary for setting up a multicast socket and should be used in a case-by-case basis instead of for all multicast sockets but otherwise I agree.